lib/charms/operator_libs_linux/v1/systemd.py

```diff
@@ -66,127 +70,65 @@
     """
-    cmd = ["systemctl", sub_cmd]
-
-    if service_name is not None:
-        cmd.append(service_name)
-    if now is not None:
-        cmd.append("--now")
-    if quiet is not None:
-        cmd.append("--quiet")
-    if sub_cmd != "is-active":
-        logger.debug("Attempting to {} '{}' with command {}.".format(cmd, service_name, cmd))
-    else:
-        logger.debug("Checking if '{}' is active".format(service_name))
-
-    proc = subprocess.Popen(cmd, **_popen_kwargs())
-    last_line = ""
-    for line in iter(proc.stdout.readline, ""):
-        last_line = line
-        logger.debug(line)
-
-    proc.wait()
-
-    if proc.returncode < 1:
-        return True
-
-    # If we are just checking whether a service is running, return True/False, rather
-    # than raising an error.
-    if sub_cmd == "is-active" and proc.returncode == 3:  # Code returned when service not active.
+    cmd = ["systemctl", *args]
+    logger.debug(f"Executing command: {cmd}")
+    proc = subprocess.run(
+        [*cmd],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        text=True,
+        bufsize=1,
+        encoding="utf-8"
+    )
+    logger.debug(proc.stdout)
+    if proc.returncode == 1 and "is-failed" in cmd:
         return False
-
-    if sub_cmd == "is-failed":
+    elif proc.returncode == 3 and "is-active" in cmd:
         return False
-
-    raise SystemdError(
-        "Could not {}{}: systemd output: {}".format(
-            sub_cmd, " {}".format(service_name) if service_name else "", last_line
-        )
-    )
-
-
-def service_running(service_name: str) -> bool:
-    """Determine whether a system service is running.
-
-    Args:
-        service_name: the name of the service to check
-    """
-    return _systemctl("is-active", service_name, quiet=True)
-
-
-def service_failed(service_name: str) -> bool:
-    """Determine whether a system service has failed.
-
-    Args:
-        service_name: the name of the service to check
-    """
-    return _systemctl("is-failed", service_name, quiet=True)
-
-
-def service_start(service_name: str) -> bool:
-    """Start a system service.
-
-    Args:
-        service_name: the name of the service to start
-    """
-    return _systemctl("start", service_name)
-
-
-def service_stop(service_name: str) -> bool:
-    """Stop a system service.
-
-    Args:
-        service_name: the name of the service to stop
-    """
-    return _systemctl("stop", service_name)
-
-
-def service_restart(service_name: str) -> bool:
-    """Restart a system service.
-
-    Args:
-        service_name: the name of the service to restart
-    """
-    return _systemctl("restart", service_name)
+    elif proc.returncode == 0:
+        return True
+    else:
+        raise SystemdError(f"Command {cmd} failed. systemctl output:\n{proc.stdout}")
+
+
+service_running = functools.partial(_systemctl, "--quiet", "is-active")
+service_running.__doc__ = "Determine whether a system service is running."
```

Comment:
So I'm okay with using functools.partial, and thanks for setting the docstring. However, I'm concerned that we're losing the type annotation here. It looks like `functools.partial` does copy across the type signature (see `inspect.signature`), but in this case `_systemctl` doesn't use type annotations, so the partial functions won't either.

Relatedly, what will the docs look like here? https://charmhub.io/operator-libs-linux/libraries/systemd#service_running ... I think we need to keep these informative (and with type annotations). I notice that even as it stands they don't have return type annotations (`bool` for some functions, `None` for others?), so maybe we can address that while we're at it.

This *may* mean writing out all the functions rather than using `functools.partial`, but I think if so, that's an okay trade-off. Better to make it nice for our users every time (docs and type annotations) rather than nice for us once.

Some of these functions also return `bool` (I don't see why funcs like `service_start` should return bool as I believe they raise an exception on failure, but we're probably locked into the existing type signatures unless we release a major version bump).

One more thing: I realize you're just copying the existing docstrings, but it's currently unclear what these first two functions do after "determining whether" the thing occurred (they return the result as a bool). I like the Go style of saying "Report whether ...", but "Return True if ..., otherwise False" is okay too.

Reply:
I can change the docstrings for `service_running(...)` and `service_failed(...)` to be more descriptive than they are now. Do not want to create any sense of ambiguity!

I believe we can type `*args` in `_systemctl(...)` by using the [Unpack](https://docs.python.org/3/library/typing.html?highlight=unpack#typing.Unpack) type provided by the `typing_extensions` module. Unfortunately, our suspicion that Charmhub will not render the partial functions in the charmlib documentation is like true, so I will need to switch them back to the tradition `def service_start(...) -> bool:` format.

Yes, I agree that it doesn't make sense for methods like `service_start(...)` and `service_stop(...)` to return a bool value, but this is also a holdover from the original implementation of the `_systemctl(...)` method where it was being used to evaluate the output of "is-active" and "is-failed". We could change it, but we will need to release a new major version rather than a patch version since I do not know if there are other charmers who depend upon this behavior.

Reply:
So I switched back to using proper function definitions rather than using `functools.partial` because functions defined using `functools.partial` will not be rendered properly in Charmhub's charmlib documentation. For `*args`, I just typed it as `str` because I did not want to bring in `typing_extensions` as a dependency. I also updated the docstrings of `service_running(...)` and `service_failed(...)` to be more descriptive.

I also changed the signatures for methods like `service_start(...)`, `service_stop(...)`, etc to return `None` rather than a bool. It makes sense to not return anything since `service_start(...)` should throw an error if a service fails to start. The charm author should consciously chose to ignore that their service has failed to start correctly by handling the raised `SystemdError`. Only thing is that we should determine if this change warrants bumping the LIBAPI version of the systemd library. Let me know what you think.

Reply:
Good changes, thanks. However, I do think we should switch back to these returning true to avoid a major version bump. See my comments below.